x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/AbstractRateAggregator.java

```diff
@@ -54,21 +57,42 @@ public AbstractRateAggregator(
         }
         this.rateUnit = rateUnit;
         this.rateMode = rateMode;
-        this.sizedBucketAggregator = findSizedBucketAncestor();
+        this.rateDivisorCalculator = new CachedSupplier<>(this::constructRateDivisor);
     }
 
-    private SizedBucketAggregator findSizedBucketAncestor() {
+    private RateDivisorCalculator constructRateDivisor() {
         SizedBucketAggregator sizedBucketAggregator = null;
+        boolean isParent = false;
         for (Aggregator ancestor = parent; ancestor != null; ancestor = ancestor.parent()) {
             if (ancestor instanceof SizedBucketAggregator) {
                 sizedBucketAggregator = (SizedBucketAggregator) ancestor;
+                isParent = ancestor == parent;
                 break;
             }
         }
         if (sizedBucketAggregator == null) {
-            throw new IllegalArgumentException("The rate aggregation can only be used inside a date histogram");
+            for (Aggregator ancestor = parent; ancestor != null; ancestor = ancestor.parent()) {
```

Comment:
Any reason to not combine this loop with the loop above?

Reply:
@imotov it depends. Its now possible to have BOTH a date_histogram and a composite agg parent that provide a sizedBucketAggregator. 

Previously, we simply grabbed the first sizedBucketAggregator we saw. If that is what we want to continue, then thats fine. I was just thinking we should prefer the `date_histogram` if we found it over the composite agg.

Reply:
Can you provide an example, that would demonstrate this issue? I am not sure I can come up with one.